airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,273 @@
+        self._policies: list[tuple[RequestMatcher, CallRatePolicy]] = []
+
+    def add_policy(self, request_matcher: RequestMatcher, policy: CallRatePolicy) -> None:
+        """Add policy for calls
+
+        :param request_matcher: callable to match request object with corresponding policy
+        :param policy: to acquire calls
+        :return:
+        """
+
+        self._policies.append((request_matcher, policy))
+
+    def acquire_call(self, request: Any, block: bool = True, timeout: Optional[float] = None) -> None:
+        """Try to get a call from budget, will block by default
+
+        :param request:
+        :param block: when true (default) will block the current thread until call credit is available
+        :param timeout: if set will limit maximum time in block, otherwise will wait until credit is available
+        :raises: CallRateLimitHit - when no credits left and if timeout was set the waiting time exceed the timeout
+        """
+
+        for matcher, policy in self._policies:
+            if matcher(request):
+                self._do_acquire(request, policy, block, timeout)
+                break
+
+    def _do_acquire(self, request: Any, policy: CallRatePolicy, block: bool, timeout: Optional[float]) -> None:
+        """Internal method to try to acquire a call credit
+
+        :param request:
+        :param policy:
+        :param block:
+        :param timeout:
+        """
+        try:
+            policy.try_acquire(request)
+        except CallRateLimitHit as exc:
+            if block:
+                if timeout is not None:
+                    time_to_wait = min(timedelta(seconds=timeout), exc.time_to_wait)
+                else:
+                    time_to_wait = exc.time_to_wait
+
+                time.sleep(time_to_wait.total_seconds())
+                policy.try_acquire(request)
+            else:
+                raise
+
+
+class SessionProxyWithCallRate:
+    def __init__(self, session: Union[requests.Session, CachedSession], api_budget: APIBudget):
+        """Wraps Session to take into account API call rate limits
+
+        :param session:
+        :param api_budget:
+        """
+        self._session = session
+        self._api_budget = api_budget
+
+    def __getattr__(self, item: str) -> Any:
+        """Forward everything to original Session class
+
+        :param item: attribute name
+        :return:
+        """
+        return getattr(self._session, item)
+
+    def __setattr__(self, key: str, value: Any) -> None:
+        """Forward everything to original Session class"""
+        return object.__setattr__(self._session, key, value)
+
+    def send(self, request: requests.PreparedRequest, **kwargs: Any) -> requests.Response:
+        """Override method to respect API call rate limits
+
+        :param request:
+        :param kwargs:
+        :return:
+        """
+        if isinstance(self._session, CachedSession):
+            if not self.cache.contains(request=request):
```

Comment:
I don't know much about threading in python but this would probably not be thread-safe in other language. The cache might contain the item when running `self.cache.contains` but might not contain the cache entry when running `self._session.send`. Is this an issue in this actual case? 

The worst consequence here is that we would exceed the rate limiting because we would by-pass the api_budget and hit the cache hence exceeding the rate limit

Reply:
The cache is threadsafe. However we encountered another issue here while testing, frequent locks of cache storage when we have large number of readers (5+).
Also the cache expiration is enabled so the case with missing records is unlikely

Reply:
my understanding from docs for the sqlite backend we're using is that writes are sequential, but it supports [unlimited concurrent reads](https://requests-cache.readthedocs.io/en/stable/user_guide/backends/sqlite.html#concurrency).

I'm separately looking into the locking issue as part of https://github.com/airbytehq/airbyte/issues/31585

Reply:
The cache is threadsafe but I fear that our usage of the cache is not. Assuming the sequence of event:

0. Entry is added to the cache
1. `if not self.cache.contains(request=request)` is executed and we don't acquire_call
2. The cache entry is cleared
3. `self._session.send` is executed and can't rely on the cache as the entry has been removed

Is this scenario possible? If not, why?

Reply:
💡 💡 💡 !

You're right. We're making an implicit assumption that the cache is never cleared. That might currently be true since requests don't expire by default, but I'm in favor of removing surprises.

@keu it seems like we can create a [rate limited cached session](https://requests-cache.readthedocs.io/en/stable/user_guide/compatibility.html#requests-ratelimiter) with [requests-ratelimiter](https://github.com/JWCook/requests-ratelimiter) which seems to do what we want (not counting cache hits against the rate limit)

```
class CachedLimiterSession(CacheMixin, LimiterMixin, Session):
    """Session class with caching and rate-limiting behavior. Accepts arguments for both
    LimiterSession and CachedSession.
    """

# Limit non-cached requests to 5 requests per second, with unlimited cached requests
# Optionally use Redis as both the bucket backend and the cache backend
session = CachedLimiterSession(
    per_second=5,
    bucket_class=RedisBucket,
    backend=RedisCache(),
)
```

Reply:
AFAIK, there are no calls to clearing of cache during the sync

Reply:
@girarda thanks, I will check it

Reply:
@maxi297 , @girarda, so I went through the code of request-ratelimiter 
and here are my conclusions:

1. the Mixin design allows us to solve with caching and remove proxy hacks, I can't say that the design based on MRO is perfect, but it is better than typing/proxy hacks and potential race conditions + unlikely some on will inherit this class.
2. Taking into account that behavior is not 100% idea we will need either another wrapper or just to copy the idea and do our own implementation (which basically just 50 lines of code). I vote for the latter. This implementation will throw a RateLimitHit exception with the time_to_wait attribute.

WDYT guys ?